// sequ/lib/index.js
'use strict';

exports.__esModule = true;

var _concatenate = require('./concatenate');

var _concatenate2 = _interopRequireDefault(_concatenate);

var _withProps = require('./withProps');

var _withProps2 = _interopRequireDefault(_withProps);

var _tillTruly = require('./tillTruly');

var _tillTruly2 = _interopRequireDefault(_tillTruly);

var _whileTruly = require('./whileTruly');

var _whileTruly2 = _interopRequireDefault(_whileTruly);

var _invoke = require('./invoke');

var _invoke2 = _interopRequireDefault(_invoke);

var _fork = require('./fork');

var _fork2 = _interopRequireDefault(_fork);

var _stub = require('./stub');

var _stub2 = _interopRequireDefault(_stub);

var _memoize = require('./memoize');

var _memoize2 = _interopRequireDefault(_memoize);

var _mapArgs = require('./mapArgs');

var _mapArgs2 = _interopRequireDefault(_mapArgs);

var _onError = require('./onError');

var _onError2 = _interopRequireDefault(_onError);

var _then = require('./then');

var _then2 = _interopRequireDefault(_then);

var _tillMatch = require('./tillMatch');

var _tillMatch2 = _interopRequireDefault(_tillMatch);

var _zipProps = require('./zipProps');

var _zipProps2 = _interopRequireDefault(_zipProps);

var _branch = require('./branch');

var _branch2 = _interopRequireDefault(_branch);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var composite = _concatenate2.default;
var sequence = _concatenate2.default;
var sequ = _concatenate2.default;
var seq = _concatenate2.default;

var main = function main() {
  return composite.apply(undefined, arguments);
};

Object.assign(main, {
  composite: composite,
  sequence: sequence,
  sequ: sequ,
  seq: seq,
  concatenate: _concatenate2.default,
  withProps: _withProps2.default,
  tillTruly: _tillTruly2.default,
  whileTruly: _whileTruly2.default,
  stub: _stub2.default,
  memoize: _memoize2.default,
  invoke: _invoke2.default,
  fork: _fork2.default,
  mapArgs: _mapArgs2.default,
  onError: _onError2.default,
  then: _then2.default,
  tillMatch: _tillMatch2.default,
  zipProps: _zipProps2.default,
  branch: _branch2.default
});

exports.default = main;
module.exports = exports['default'];
